apps/ejabberd/src/mod_muc_db_mnesia.erl

```diff
@@ -0,0 +1,239 @@
+init(Host, Opts) ->
+    mnesia:create_table(muc_room,
+                        [{disc_copies, [node()]},
+                         {attributes, record_info(fields, muc_room)}]),
+    mnesia:create_table(muc_registered,
+                        [{disc_copies, [node()]},
+                         {attributes, record_info(fields, muc_registered)}]),
+    mnesia:add_table_copy(muc_room, node(), disc_copies),
+    mnesia:add_table_copy(muc_registered, node(), disc_copies),
+    MyHost = gen_mod:get_opt_subhost(Host, Opts, <<"conference.@HOST@">>),
+    update_tables(MyHost),
+    mnesia:add_table_index(muc_registered, nick),
+    ok.
+
+-spec store_room(ejabberd:server(), ejabberd:server(), mod_muc:room(), list())
+            -> {'aborted',_} | {'atomic',_}.
+store_room(_LServer, Host, Name, Opts) ->
+    F = fun() ->
+                mnesia:write(#muc_room{name_host = {Name, Host},
+                                       opts = Opts})
+        end,
+    mnesia:transaction(F).
+
+
+-spec restore_room(ejabberd:server(), ejabberd:server(), mod_muc:room())
+                                    -> 'error' | 'undefined' | [any()].
+restore_room(_LServer, Host, Name) ->
+    case catch mnesia:dirty_read(muc_room, {Name, Host}) of
+        [#muc_room{opts = Opts}] ->
+            Opts;
+        _ ->
+            error
+    end.
+
+-spec forget_room(ejabberd:server(), ejabberd:server(), mod_muc:room()) -> 'ok'.
+forget_room(_LServer, Host, Name) ->
+    F = fun() ->
+                mnesia:delete({muc_room, {Name, Host}})
+        end,
+    mnesia:transaction(F),
+    ejabberd_hooks:run(forget_room, Host, [Host, Name]),
+    ok.
+
+get_rooms(_Lserver, Host) ->
+    mnesia:dirty_select(
+                 muc_room, [{#muc_room{name_host = {'_', Host}, _ = '_'},
+                             [],
+                             ['$_']}]).
+
+-spec can_use_nick(ejabberd:server(), ejabberd:server(),
+                   ejabberd:jid(), mod_muc:nick()) -> boolean().
+can_use_nick(_LServer, Host, JID, Nick) ->
+    {LUser, LServer, _} = jid:to_lower(JID),
+    LUS = {LUser, LServer},
+    case catch mnesia:dirty_select(
+                 muc_registered,
+                 [{#muc_registered{us_host = '$1',
+                                   nick = Nick,
+                                   _ = '_'},
+                   [{'==', {element, 2, '$1'}, Host}],
+                   ['$_']}]) of
+        {'EXIT', _Reason} ->
+            true;
+        [] ->
+            true;
+        [#muc_registered{us_host = {U, _Host}}] ->
+            U == LUS
+    end.
+
+get_nick(_LServer, Host, From) ->
+    {LUser, LServer, _} = jid:to_lower(From),
+    LUS = {LUser, LServer},
+    case mnesia:dirty_read(muc_registered, {LUS, Host}) of
+    [] -> error;
+    [#muc_registered{nick = Nick}] -> Nick
+    end.
+
+set_nick(_LServer, Host, From, Nick) ->
+    {LUser, LServer, _} = jid:to_lower(From),
+    LUS = {LUser, LServer},
+    F = fun () ->
+        case Nick of
+            <<"">> ->
```

Comment:
I think it is a good PR to make this API better and actually have dedicated function for unregistering the nick.